x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -419,16 +418,24 @@ public void restoreShard(
             });
             // TODO: There should be some local timeout. And if the remote cluster returns an unknown session
             // response, we should be able to retry by creating a new session.
-            final RestoreSession restoreSession = openSession(metadata.name(), remoteClient, leaderShardId, shardId, recoveryState);
-            toClose.addFirst(restoreSession); // Some tests depend on closing session before cancelling retention lease renewal
-            restoreSession.restoreFiles(store, restoreListener.delegateFailureAndWrap((l, v) -> {
-                logger.trace("[{}] completed CCR restore", shardId);
-                updateMappings(remoteClient, leaderIndex, restoreSession.mappingVersion, client, shardId.getIndex());
-                l.onResponse(null);
-            }));
-        } catch (Exception e) {
-            restoreListener.onFailure(e);
-        }
+            ActionListener<RestoreSession> sessionListener = restoreListener1.delegateFailureAndWrap(
+                // Some tests depend on closing session before cancelling retention lease renewal.
+                (l1, restoreSession) -> restoreSession.restoreFiles(store, new ActionListener<>() {
+                    @Override
+                    public void onResponse(Void unused) {
+                        logger.trace("[{}] completed CCR restore", shardId);
+                        updateMappings(remoteClient, leaderIndex, restoreSession.mappingVersion, client, shardId.getIndex());
+                        restoreSession.close(l1);
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        restoreSession.close(ActionListener.runAfter(ActionListener.noop(), () -> l1.onFailure(e)));
```

Comment:
This is equivalent I think?

```suggestion
                        restoreSession.close(ActionListener.running(() -> l1.onFailure(e)));
```